lib/Reader/abisignature.cpp

```diff
@@ -553,23 +708,41 @@ Function *ABIMethodSignature::createFunction(GenIR &Reader, Module &M) {
       I++;
     }
 
-    if (Arg.getKind() == ABIArgInfo::Indirect) {
-      // TODO: byval attribute support
+    switch (Arg.getKind()) {
+    case ABIArgInfo::Indirect:
       ArgumentTypes[I] = Reader.getManagedPointerType(Arg.getType());
-    } else {
-      ArgumentTypes[I] = Arg.getType();
+      break;
 
-      if (Arg.getKind() == ABIArgInfo::ZeroExtend) {
-        ArgAttrs.addAttribute(Attribute::ZExt);
-      } else if (Arg.getKind() == ABIArgInfo::SignExtend) {
-        ArgAttrs.addAttribute(Attribute::SExt);
+    case ABIArgInfo::Expand:
+      for (const ABIArgInfo::Expansion &Exp : Arg.getExpansions()) {
+        ArgumentTypes[I++] = Exp.TheType;
+      }
+      I--;
+      break;
+
+    case ABIArgInfo::ZeroExtend:
+      ArgAttrs.addAttribute(Attribute::ZExt);
+      goto direct;
+
+    case ABIArgInfo::SignExtend:
+      ArgAttrs.addAttribute(Attribute::SExt);
+      goto direct;
+
+    case ABIArgInfo::Direct: {
+    direct:
+      Type *ArgTy = Arg.getType();
+      if (ArgTy->isStructTy()) {
+        ArgTy = Reader.getManagedPointerType(ArgTy);
```

Comment:
why managed?  I'd have expected that
 1. this will always be something allocated in the caller's frame, so outside the GC heap, and
 2. we won't be inserting write-barriers for stores into it

if that's right, then unmanaged would seem to make more sense.